### ISSUE TITLE
`linalg` routines fail with pathologically edge cases

### ISSUE THREAD

**Issue by leofang** [cupy/cupy#6160]
### Description

This is discovered using the `array-api-tests`. I am not sure if the same behavior for handling 2D "scalar" matrices is applicable to all `numpy.linalg` routines, or just the two I posted here.

### To Reproduce

```py
>>> import numpy as np
>>> import cupy as cp
>>>
>>> # cholesky
>>> np.linalg.cholesky(np.empty((0, 0), dtype=np.float32))
array([], shape=(0, 0), dtype=float32)
>>> cp.linalg.cholesky(cp.empty((0, 0), dtype=cp.float32))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/leof/dev/cupy_cuda114/cupy/linalg/_decomposition.py", line 205, in cholesky
    potrf(
  File "cupy_backends/cuda/libs/cusolver.pyx", line 1205, in cupy_backends.cuda.libs.cusolver.spotrf
    cpdef spotrf(intptr_t handle, int uplo, int n, size_t A, int lda,
  File "cupy_backends/cuda/libs/cusolver.pyx", line 1212, in cupy_backends.cuda.libs.cusolver.spotrf
    check_status(status)
  File "cupy_backends/cuda/libs/cusolver.pyx", line 1062, in cupy_backends.cuda.libs.cusolver.check_status
    raise CUSOLVERError(status)
cupy_backends.cuda.libs.cusolver.CUSOLVERError: CUSOLVER_STATUS_INVALID_VALUE
>>>
>>> # matrix_power
>>> np.linalg.matrix_power(np.empty((0, 0), dtype=np.float32), n=-1)
array([], shape=(0, 0), dtype=float32)
>>> cp.linalg.matrix_power(cp.empty((0, 0), dtype=cp.float32), n=-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/leof/dev/cupy_cuda114/cupy/linalg/_product.py", line 374, in matrix_power
    M = _solve.inv(M)
  File "/home/leof/dev/cupy_cuda114/cupy/linalg/_solve.py", line 323, in inv
    cupyx.lapack.gesv(a, b)
  File "/home/leof/dev/cupy_cuda114/cupyx/lapack.py", line 68, in gesv
    getrf(handle, n, n, a.data.ptr, n, dwork.data.ptr, dipiv.data.ptr,
  File "cupy_backends/cuda/libs/cusolver.pyx", line 1402, in cupy_backends.cuda.libs.cusolver.sgetrf
    cpdef sgetrf(intptr_t handle, int m, int n, size_t A, int lda,
  File "cupy_backends/cuda/libs/cusolver.pyx", line 1409, in cupy_backends.cuda.libs.cusolver.sgetrf
    check_status(status)
  File "cupy_backends/cuda/libs/cusolver.pyx", line 1062, in cupy_backends.cuda.libs.cusolver.check_status
    raise CUSOLVERError(status)
cupy_backends.cuda.libs.cusolver.CUSOLVERError: CUSOLVER_STATUS_INVALID_VALUE
```

**Comment by toslunar** [comment]
The previous tracking issue is #4330.

**Comment by leofang** [comment]
Thanks for the fixes @toslunar, looks like this issue can be closed now? 🙂

**Comment by toslunar** [comment]
I'll fix `eigh` and `eigvalsh`.

**Comment by toslunar** [comment]
Fixed by #6164, #6167, #6210.

**Comment by leofang** [comment]
Many thanks, @toslunar!
